## 2η Σειρά Εργαστηριακών Ασκήσεων

### Όνομα: Αλεξανδρόπουλος Σταμάτης
### Αριθμός Μητρώου: 03117060
### Σχολή: ΗΜΜΥ

____

### Άσκηση 19

Γράψτε σε ένα καινούργιο Jupyter Notebook έναν ολοκληρωμένο κώδικα σε Python που θα κάνει $M=50$ εκτιμήσεις Monte Carlo της πιθανότητας $\mathbb{P}_1[X_{20}=1]$ για την αλυσίδα της **Άσκησης 19** και θα υπολογίζει τη δειγματική μέση τιμή και διασπορά αυτών των εκτιμήσεων. Σε ένα κελί κειμένου παρακάτω απαντήστε τις εξής ερωτήσεις.

1. Ποια είναι ο μέσος όρος των $Μ$ εκτιμήσεων της πιθανότητας $\mathbb{P}_1[X_{20}=1]$ όταν $N=200$ και ποιος όταν $N=20.000$;
2. Είναι αυτά τα αποτελέσματα κοντά στη θεωρητική τιμή που έχετε υπολογίσει;
3. Ποια είναι η δειγματική διασπορά των $M$ εκτιμήσεων στις δύο περιπτώσεις;
4. Αλλάζει ουσιαστικά η εκτίμησή σας, αν η αλυσίδα ξεκινάει από την κατάσταση `3` (τρίτη γραμμή του πίνακα);

In [1]:
import random 
import statistics as stat
import random
random.seed(2021)  # for reproducibility
from simple_markov_chain_lib import markov_chain


M=50
steps=20
list_prob=[1/6,1/12]
for p in list_prob:
    if (p==1/6):
        print("Όταν η πιθανότητα είναι 1/6 ")
    else:
        print("Όταν η πιθανότητα είναι 1/12")
    initial_list=[1,3]
    for initial in [1,3]:
        initial_probs={initial: 1.0}
        print("Όταν η αλυσίδα ξεκινάει από την κατάσταση ",initial)
        markov_table = {
        1: {2: 1.},
        2: {2: 2/3, 3: 1/3},
        3: {1: p, 2: 1-p}
        }
        mc = markov_chain(markov_table, initial_probs)
        list2=[200,20000]
        for N in list2:
            M=50
            steps=20
            list_track=[]
            for i in range(M):
                counter = 0
                for i in range(N): 
                    mc.start()
                    for j in range(steps):mc.move()
                    if mc.running_state == 1:  counter += 1
                list_track.append(counter/N)
            print(
                """Ο μέσος όρος Μ=50 εκτιμήσεων για Ν={0} είναι {1} και η διασπορά είναι {2}""".format(N,round(stat.mean(list_track),6),round(stat.variance(list_track),7))
            )

    


Όταν η πιθανότητα είναι 1/6 
Όταν η αλυσίδα ξεκινάει από την κατάσταση  1
Ο μέσος όρος Μ=50 εκτιμήσεων για Ν=200 είναι 0.0439 και η διασπορά είναι 0.0002034
Ο μέσος όρος Μ=50 εκτιμήσεων για Ν=20000 είναι 0.040159 και η διασπορά είναι 1.7e-06
Όταν η αλυσίδα ξεκινάει από την κατάσταση  3
Ο μέσος όρος Μ=50 εκτιμήσεων για Ν=200 είναι 0.036 και η διασπορά είναι 0.000198
Ο μέσος όρος Μ=50 εκτιμήσεων για Ν=20000 είναι 0.040341 και η διασπορά είναι 1.8e-06
Όταν η πιθανότητα είναι 1/12
Όταν η αλυσίδα ξεκινάει από την κατάσταση  1
Ο μέσος όρος Μ=50 εκτιμήσεων για Ν=200 είναι 0.0202 και η διασπορά είναι 0.0001112
Ο μέσος όρος Μ=50 εκτιμήσεων για Ν=20000 είναι 0.020422 και η διασπορά είναι 8e-07
Όταν η αλυσίδα ξεκινάει από την κατάσταση  3
Ο μέσος όρος Μ=50 εκτιμήσεων για Ν=200 είναι 0.0213 και η διασπορά είναι 0.0001233
Ο μέσος όρος Μ=50 εκτιμήσεων για Ν=20000 είναι 0.020581 και η διασπορά είναι 9e-07


#### Απαντήσεις 
Τα ερωτήματα 1 και 3 έχουν απαντηθεί παραπάνω.

2)O υπολογίσμός της θεωρητικής μέσης τιμής φαίνεται παρακάτω. Ουσιαστικά υψώνουμε τον πίνακα πιθανοτήτων μετάβασης στην 20η δύναμη και παίρνουμε την τιμή του κελιού (0,0) για τις αντιστοίχες πιθανότητες p=1/6 και p=1/12. 

Παρατηρούμε ότι όντως οι τιμές που λαμβάνουμε όπως φαίνεται παρακάτω είναι πολύ κοντά στις εκτιμήσεις που βρήκαμε με τη μέθοδο Monte Carlo, αφού στην περίπτωση που :
- p=1/6 :  Pr[X_20 = 1 | X_0 = 1]  =0.0439
- p=1/6 :  Pr[X_20 = 1 | X_0 = 1]  =0.0202


In [3]:
import numpy as np
for p in [1/6,1/12]:
  arr = np.array([[0,1,0],[0,2/3,1/3],[p,1-p,0]])
  for n in range(20):
        arr = arr.dot(arr)
  print (arr[0][0])

0.03999999999951925
0.02040816326507903


4) Παρατηρούμε οτι όταν η αλυσίδα ξεκινά από την τιμή 3 οι εκτιμήσεις μας είναι σχεδόν ίδιες με αυτές που έχουμε όταν η αλυσίδα ξεκινά από την τιμή 1.

### Άσκηση 40


Υπολογίστε προσεγγιστικά με Monte Carlo την πιθανότητα νίκης του παίκτη που σερβίρει (Player A), αν η πιθανότητα που έχει να κερδίσει κάθε πόντο είναι 𝑝=0,6.

![Markov Chain of Tennis Game](https://www.wolfram.com/mathematica/new-in-9/markov-chains-and-queues/HTMLImages.en/analyze-a-tennis-game/O_23.png)

In [6]:
import random
import numpy as np
import statistics as stat
from simple_markov_chain_lib import markov_chain
random.seed(2021)  # for reproducibility

p = 0.6
q=1-p
initial_probs = {"0-0": 1.0}
markov_table = {
    "0-0": { "15-0": p,"0-15": q},
    "15-0": { "30-0": p,"15-15":q},
    "0-15": {"15-15": p,"0-30": q },
    "30-0": { "40-0": p,"30-15": q },
    "15-15": { "30-15": p,"15-30": q },
    "0-30": { "15-30": p,"15-40": q },
    "40-0": { "GameA": p,"40-15": q },
    "30-15": { "40-15": p,"Deuce": q },
    "15-30": { "Deuce": p,"15-40": q },
    "0-40": { "15-40": p,"GameB": q },
    "40-15": { "GameA": p,"AdvA": q },
    "Deuce": { "AdvA": p,"AdvB": q },
    "15-40": { "AdvB": p,"GameB": q},
    "GameA": { "GameA": 1.0 },
    "AdvA": { "GameA": p,"Deuce": q },
    "AdvB": { "Deuce": p,"GameB": q},
    "GameB": { "GameB": 1.0 },
}

mc = markov_chain( markov_table, initial_probs)

#Experiment 
N=20000
list_track=[]
count=0
for i in range(10):
    for i in range(N):
        mc.start() 
        while mc.running_state not in ["GameA","GameB"]:
            mc.move()
        if mc.running_state=="GameA":
            count += 1
    mean=count/N
    list_track.append(mean)
    count=0
result= np.mean(list_track)
print('Αρα υπολογίσαμε προσεγγιστικά ότι η πιθανότητα νίκης του παίκτη Player A είναι {0:.5f}'.format(result))

Αρα υπολογίσαμε προσεγγιστικά ότι η πιθανότητα νίκης του παίκτη Player A είναι 0.74146
